src/Cli/dotnet/commands/dotnet-sln/add/Program.cs

```diff
@@ -100,9 +102,9 @@ private async Task AddProjectsToSolutionAsync(string solutionFileFullPath, IEnum
             foreach (var projectPath in projectPaths)
             {
                 string relativePath = Path.GetRelativePath(Path.GetDirectoryName(solutionFileFullPath), projectPath);
-                // Add fallback solution folder if relative path does not contain `..`.
-                string relativeSolutionFolder =  relativePath.Split(Path.DirectorySeparatorChar).Any(p => p == "..")
-                    ? string.Empty : Path.GetDirectoryName(relativePath);
+
+                // Add fallback solution folder if relative path does not contain invalid characters
+                string relativeSolutionFolder = DoesSolutionFolderPathContainInvalidChars(relativePath) ? string.Empty : Path.GetDirectoryName(relativePath);
```

Comment:
I got turned around twice reading this. I think I understand it properly now. I feel like there has to be a better way to express this, but I haven't figured one out yet, to be honest

Reply:
Yes, It is very messy. I am refactoring the whole "automatically generate intermediate solution folders" part to be more clear on what it should do